ADD Show diff between the expected output and the actual output in the autograder

PIG208 opened the pull request:
The plan is to add a column called `diff` to `SubmissionTestResult`, which is usually `None`.

I'm adding a column instead of a special directive that can be parsed when it shows up in stdout. Because we will have to implement and maintain an extra parser in the frontend, while having a separate field allows us to feed it to `react-diff-view` directly.

Fixes #254

wabscale commented:
Hey you should hold off on this. This is going to be a piece of much larger changes to the way that autograding happens. We're going to be doing this during winter break.

PIG208 commented:
I will do a quick push tmr to kind of see how this turns out.

PIG208 commented:
OK, it seems to be a bit more complicated than I expected. Because now I'm trying to set up test cases before actually making the changes. So it might take longer.

PIG208 commented:
![image](https://user-images.githubusercontent.com/39874143/147844770-83c3f06a-0813-439c-86cd-826d0a46c850.png)
![image](https://user-images.githubusercontent.com/39874143/147844773-361d0783-501f-43ec-b4ad-56458632aaf7.png)
![image](https://user-images.githubusercontent.com/39874143/147844808-caadd1df-3644-44db-ae26-549b02a2351b.png)

PIG208 commented:
I just updated the PR. Specifically, I reworked `test_lines` in bf3f1e5b359a982974582019aff73988167551ba. Previously we just preprocess all the lines and compute the diff of the entire outputs. This improves the performance by only parsing the diff within a limited context where we have the first mismatch.

wabscale commented:
I'm going to get on later and push some changes before rebasing this. I need to adjust a few things so that it will fit into the changes to the autograde tests format.
